include/swift/AST/TypeCheckRequests.h

```diff
@@ -322,6 +322,71 @@ class USRGenerationRequest :
   bool isCached() const { return true; }
 };
 
+/// Needed to use a DeclContext* as a request parameter.
+/// Without this wrapper, compilation fails missing simple_display for
+/// DeclContext* Defining said simple_display results in ambiguitites when
+/// simple_display is called for some Decl subclass pointers.
+
+struct DeclContextWrapper {
+  const DeclContext *dc;
+
+  DeclContextWrapper(const DeclContext *dc) : dc(dc) {}
+
+  SourceLoc getLoc() const { return SourceLoc(); }
+
+  friend hash_code hash_value(const DeclContextWrapper &x) {
+    return hash_value(x.dc);
+  }
+
+  friend bool operator==(const DeclContextWrapper &lhs,
+                         const DeclContextWrapper &rhs) {
+    return lhs.dc == rhs.dc;
+  }
+
+  friend bool operator!=(const DeclContextWrapper &lhs,
+                         const DeclContextWrapper &rhs) {
+    return !(lhs == rhs);
+  }
+};
+
+void simple_display(llvm::raw_ostream &out, const DeclContextWrapper &x);
+class TypeChecker;
+
+// Find the type in the cache or look it up
+class DefaultTypeRequest
+    : public SimpleRequest<DefaultTypeRequest, CacheKind::SeparatelyCached,
+                           Type, const char *, bool, const DeclContextWrapper> {
```

Comment:
Same comment about  the `const char *` here: we'd like something that either owns string data or doesn't require ownership.

Reply:
I like an `enum` better on performance grounds and clarity, too.

Reply:
Redone using the existing enum.

Reply:
Great!

Reply:
:)